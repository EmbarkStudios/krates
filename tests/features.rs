mod util;

/// Ensures weak dependencies are properly pruned if not explicitly pulled in
/// https://github.com/EmbarkStudios/krates/issues/41
#[test]
fn prunes_multiple_weak_features() {
    let mut cmd = krates::Cmd::new();
    cmd.manifest_path("tests/features/Cargo.toml")
        .features(
            ["blocking", "json", "multipart", "stream"]
                .into_iter()
                .map(|f| f.to_owned()),
        )
        .no_default_features();

    let mut builder = krates::Builder::new();
    builder.include_targets([("x86_64-unknown-linux-gnu", vec![])]);
    let md: krates::Krates = builder.build(cmd, krates::NoneFilter).unwrap();

    for name in md.krates().map(|k| &k.name) {
        println!("{name}");
    }

    // All of the specified features have a weak dependency on reqwest, so it
    // shouldn't be present in the graph
    assert_eq!(0, md.krates_by_name("reqwest").count());
    assert_eq!(0, md.krates_by_name("reqest").count());
}

/// While the zlib features brings in git2, the openssl dependency for both
/// git2 and git2-sys is optional and weak, and since we've not explicitly
/// enabled a feature to bring it in, it should not be present in the graph,
/// even though `cargo metadata` will list it in the graph
#[test]
fn prunes_mixed_dependencies() {
    let mut cmd = krates::Cmd::new();
    cmd.manifest_path("tests/features/Cargo.toml")
        .features(["zlib".to_owned()]);

    let mut builder = krates::Builder::new();
    builder.include_targets([("x86_64-unknown-linux-gnu", vec![])]);
    let md: krates::Krates = builder.build(cmd, krates::NoneFilter).unwrap();

    assert_eq!(0, md.krates_by_name("openssl-sys").count());
    // cmake is brought in via the zlib-ng-compat feature. gross.
    assert_eq!(1, md.krates_by_name("cmake").count());
}

macro_rules! assert_features {
    ($graph:expr, $name:expr, $features:expr) => {
        let (_, krate) = $graph.krates_by_name($name).next().unwrap();

        assert_eq!($graph.get_enabled_features(&krate.id).unwrap(), $features);
    };
}

/// Ensures we can enable crate features even when that crate has been renamed
/// as well as features being precise
#[test]
fn handles_features_for_renamed_crates() {
    let mut cmd = krates::Cmd::new();
    cmd.manifest_path("tests/features/Cargo.toml")
        .no_default_features()
        .features(["midi".to_owned()]);

    let mut builder = krates::Builder::new();
    builder.include_targets([("aarch64-apple-darwin", vec![])]);
    let md: krates::Krates = builder.build(cmd, krates::NoneFilter).unwrap();

    assert_eq!(1, md.krates_by_name("coreaudio-rs").count());

    // Ensure only the single feature is enabled, not all of them
    assert_features!(md, "coreaudio-sys", ["core_midi"]);
}

/// Ensures that explicitly toggling on an optional dependency works
#[test]
fn handles_explicit_weak_features() {
    {
        let mut cmd = krates::Cmd::new();
        cmd.manifest_path("tests/features/Cargo.toml")
            .no_default_features()
            .features(["reqest".to_owned()]);

        let mut builder = krates::Builder::new();
        builder.include_targets([("x86_64-unknown-linux-musl", vec![])]);
        let md: util::Graph = builder.build(cmd, krates::NoneFilter).unwrap();

        let dg = krates::petgraph::dot::Dot::new(md.graph()).to_string();

        std::fs::write("ohno.dot", dg).unwrap();

        // assert_features!(
        //     md,
        //     "reqwest",
        //     [
        //         "__rustls",
        //         "__tls",
        //         "async-compression",
        //         "brotli",
        //         "cookie_crate",
        //         "cookie_store",
        //         "cookies",
        //         "hyper-rustls",
        //         "proc-macro-hack",
        //         "rustls",
        //         "rustls-pemfile",
        //         "rustls-tls",
        //         "rustls-tls-webpki-roots",
        //         "tokio-rustls",
        //         "tokio-util",
        //         "webpki-roots"
        //     ]
        // );
    }

    // {
    //     let mut cmd = krates::Cmd::new();
    //     cmd.manifest_path("tests/features/Cargo.toml")
    //         .no_default_features()
    //         .features(["reqest".to_owned()]);

    //     let mut builder = krates::Builder::new();
    //     builder.include_targets([("x86_64-unknown-linux-musl", vec![])]);
    //     builder.ignore_kind(krates::DepKind::Normal, krates::Scope::All);
    //     let md: krates::Krates = builder.build(cmd, krates::NoneFilter).unwrap();

    //     assert_features!(
    //         md,
    //         "reqwest",
    //         ["cookie_crate", "cookie_store", "cookies", "proc-macro-hack"]
    //     );
    // }

    // let mut cmd = krates::Cmd::new();
    // cmd.manifest_path("tests/features/Cargo.toml")
    //     .features(["reqest".to_owned()]);

    // let mut builder = krates::Builder::new();
    // builder.include_targets([("x86_64-unknown-linux-musl", vec![])]);
    // let md: krates::Krates = builder.build(cmd, krates::NoneFilter).unwrap();

    // assert_features!(md, "reqwest", ["json", "cookies"]);
}
