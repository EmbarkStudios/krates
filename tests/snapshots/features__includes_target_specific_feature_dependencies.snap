---
source: tests/features.rs
expression: dotgraph
---
digraph {
    0 [ label = "crate features-2 0.1.0 path+file:///krates/tests/features-2" ]
    1 [ label = "crate proc-macro2 1.0.79" ]
    2 [ label = "crate quote 1.0.36" ]
    3 [ label = "crate serde 1.0.197" ]
    4 [ label = "crate serde_derive 1.0.197" ]
    5 [ label = "crate smallvec 1.13.2" ]
    6 [ label = "crate syn 2.0.58" ]
    7 [ label = "crate unicode-ident 1.0.12" ]
    8 [ label = "feature serde" ]
    9 [ label = "feature proc-macro" ]
    10 [ label = "feature default" ]
    11 [ label = "feature proc-macro" ]
    12 [ label = "feature clone-impls" ]
    13 [ label = "feature derive" ]
    14 [ label = "feature parsing" ]
    15 [ label = "feature printing" ]
    16 [ label = "feature proc-macro" ]
    0 -> 5 [ label = "" ]
    0 -> 8 [ label = " 'cfg(target_os = \"android\")'" ]
    1 -> 7 [ label = "" ]
    2 -> 1 [ label = "" ]
    2 -> 9 [ label = "" ]
    3 -> 10 [ label = " 'cfg(any())'" ]
    4 -> 9 [ label = "" ]
    4 -> 11 [ label = "" ]
    4 -> 12 [ label = "" ]
    4 -> 13 [ label = "" ]
    4 -> 14 [ label = "" ]
    4 -> 15 [ label = "" ]
    4 -> 16 [ label = "" ]
    5 -> 3 [ label = "" ]
    6 -> 1 [ label = "" ]
    6 -> 9 [ label = "" ]
    6 -> 2 [ label = "" ]
    6 -> 11 [ label = "" ]
    6 -> 7 [ label = "" ]
    9 -> 1 [ label = "" ]
    11 -> 2 [ label = "" ]
    11 -> 9 [ label = "" ]
    10 -> 4 [ label = "" ]
    8 -> 5 [ label = "" ]
    8 -> 3 [ label = "" ]
    16 -> 6 [ label = "" ]
    16 -> 9 [ label = "" ]
    15 -> 6 [ label = "" ]
    15 -> 2 [ label = "" ]
    14 -> 6 [ label = "" ]
    13 -> 6 [ label = "" ]
    12 -> 6 [ label = "" ]
}
